contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java

```diff
@@ -0,0 +1,240 @@
+   */
+  private OkHttpClient setupHttpClient() {
+    Builder builder = new OkHttpClient.Builder();
+
+    // Set up the HTTP Cache.   Future possibilities include making the cache size and retention configurable but
+    // right now it is on or off.  The writer will write to the Drill temp directory if it is accessible and
+    // output a warning if not.
+    if (config.cacheResults()) {
+      setupCache(builder);
+    }
+
+    // If the API uses basic authentication add the authentication code.
+    if (apiConfig.authType().toLowerCase().equals("basic")) {
+      logger.debug("Adding Interceptor");
+      builder.addInterceptor(new BasicAuthInterceptor(apiConfig.userName(), apiConfig.password()));
+    }
+
+    // Set timeout
+    builder.connectTimeout(config.timeout(), TimeUnit.SECONDS);
+    builder.writeTimeout(config.timeout(), TimeUnit.SECONDS);
+    builder.readTimeout(config.timeout(), TimeUnit.SECONDS);
+
+    return builder.build();
+  }
+
+  /**
+   * This function accepts a Builder object as input and configures response caching. In order for
+   * caching to work, the DRILL_TMP_DIR variable must be set either as a system environment variable or in the
+   * Drill configurations.
+   * <p>
+   * The function will attempt to get the DRILL_TMP_DIR from these places, and if it cannot, it will issue a warning in the logger.
+   *
+   * @param builder Builder the Builder object to which the cacheing is to be configured
+   */
+  private void setupCache(Builder builder) {
+    int cacheSize = 10 * 1024 * 1024;   // TODO Add cache size in MB to config
+    String drillTempDir;
+
+    try {
+      if (context.getOptions().getOption(ExecConstants.DRILL_TMP_DIR) != null) {
+        drillTempDir = context.getOptions().getOption(ExecConstants.DRILL_TMP_DIR).string_val;
+      } else {
+        drillTempDir = System.getenv("DRILL_TMP_DIR");
+      }
+      File cacheDirectory = new File(drillTempDir);
+      if (cacheDirectory == null) {
+        logger.warn("HTTP Storage plugin caching requires the DRILL_TMP_DIR to be configured. Please either set DRILL_TMP_DIR or disable HTTP caching.");
+      } else {
+        Cache cache = new Cache(cacheDirectory, cacheSize);
+        logger.debug("Caching HTTP Query Results at: {}", drillTempDir);
+
+        builder.cache(cache);
+      }
+    } catch (Exception e) {
+      logger.warn("HTTP Storage plugin caching requires the DRILL_TMP_DIR to be configured. Please either set DRILL_TMP_DIR or disable HTTP caching.");
+    }
+  }
+
+  /**
+   * This function accepts text from a post body in the format:
+   * key1=value1
+   * key2=value2
+   *
+   * and creates the appropriate headers.
+   *
+   * @return FormBodu.Builder The populated formbody builder
+   */
+  private FormBody.Builder buildPostBody() {
+    final Pattern postBpdyPattern = Pattern.compile("^.+=.+$");
+
+    FormBody.Builder formBodyBuilder = new FormBody.Builder();
+    String[] lines = apiConfig.postBody().split("\\r?\\n");

```

Comment:
This code seems to assume that the post body is a multi-line text field. This is awkward in JSON, I think it needs a embedded object quote. Will the Drill web console user know how to specify that? Does Jackson do a good job of encoding/decoding this text? (Asking because I've never tried.)

We then split the body into lines, and split at an equals. So, we are assuming a URL-encoded post body? I thought some REST APIs use JSON-encoded bodies? (The so-called AJAX format, I think.)

If the body is URL-encoded, should we perhaps store it as an object in the config JSON; as name/value pairs?

Reply:
This is a good point.  In the docs, I required that the post body be in the format below:
```
key1=value1
key2=value2
```
There is a unit test for this, and it did work.  I also tried this via the web UI and didn't have any issues. 
You are correct in that the spec for HTTP POST can also accept JSON as input, however, for the initial version of this, I decided to limit it to KV pairs.  The library that I'm using (OkHTTP3) handles the URL encoding so the user does not have to worry about that.

I added a statement about this limitation in the `README.md` file.